XenAdmin/SettingsPanels/GpuEditPage.cs

```diff
@@ -322,38 +250,119 @@ public void ShowHideWarnings()
 
                 imgHA.Visible = labelHA.Visible = true;
 
-                labelGpuType.Enabled = comboBoxGpus.Enabled = false;
+                addButton.Enabled = deleteButton.Enabled = false;
                 return;
             }
 
-            labelGpuType.Enabled = comboBoxGpus.Enabled = true;
-            GpuTuple tuple = comboBoxGpus.SelectedItem as GpuTuple;
+            addButton.Enabled = !VGpus.Any(v => (Connection.Resolve(v.type)).compatible_types_in_vm.Count() == 0);
+            deleteButton.Enabled = gpuGrid.SelectedRows.Count > 0;
 
             imgStopVM.Visible = labelStopVM.Visible =
             imgHA.Visible = labelHA.Visible = false;
 
             imgRDP.Visible = labelRDP.Visible =
-                HasChanged && currentGpuTuple.GpuGroup == null &&
-                tuple != null && !tuple.IsFractionalVgpu;
+                HasChanged && currentGpus.Count == 0 &&
+                vGpus.Count > 0 && vGpus.Any(v => v.IsPassthrough());
 
             imgNeedGpu.Visible = labelNeedGpu.Visible =
-            labelNeedDriver.Visible = imgNeedDriver.Visible =
-                tuple != null && tuple.GpuGroup != null;
+                labelNeedDriver.Visible = imgNeedDriver.Visible = vGpus.Count > 0;
         }
 
-        private void comboBoxGpus_SelectedIndexChanged(object sender, EventArgs e)
+        private void warningsTable_SizeChanged(object sender, EventArgs e)
         {
+            int[] columnsWidth = warningsTable.GetColumnWidths();
+            int textColumnWidth = columnsWidth.Length > 1 ? columnsWidth[1] : 1;
+
+            labelRDP.MaximumSize = labelStopVM.MaximumSize = new Size(textColumnWidth, 999);
+        }
+
+        private void addButton_Click(object sender, EventArgs e)
+        {
+            AddVGPUDialog dialog = new AddVGPUDialog(vm, VGpus);

```

Comment:
Our recommended practice is to put the dialog into a "using" statement

Reply:
Thanks. Great to learn.